### ISSUE TITLE
Why set pad tokens to 0

### ISSUE THREAD

**Issue by Chios-C** [huggingface/trl#2235]
What is the purpose of this operation? 

`input_ids = torch.masked_fill(query_responses, ~attention_mask, 0)`

`def forward(
    model: torch.nn.Module,
    query_responses: torch.Tensor,
    pad_token_id: int,
) -> torch.nn.Module:

    attention_mask = query_responses != pad_token_id
    position_ids = attention_mask.cumsum(1) - attention_mask.long()
    input_ids = torch.masked_fill(query_responses, ~attention_mask, 0)
    return model(
        input_ids=input_ids,
        attention_mask=attention_mask,
        position_ids=position_ids,
        return_dict=True,
        output_hidden_states=True,
    )`

**Comment by qgallouedec** [comment]
This operation replaces tokens outside the attention mask with token 0. This operation has no influence on model output within the attention mask:

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

model_id = "Qwen/Qwen2.5-0.5B-Instruct"

model = AutoModelForCausalLM.from_pretrained(model_id)
tokenizer = AutoTokenizer.from_pretrained(model_id)
pad_token_id = tokenizer.pad_token_id
input_ids = torch.tensor([[pad_token_id, pad_token_id, 1, 2, 3, 4, 5, pad_token_id]])
attention_mask = input_ids != pad_token_id  # [[False, False, True, True, True, True, True, False]]
position_ids = attention_mask.cumsum(1) - attention_mask.long()  # [[0, 0, 1, 2, 3, 4, 5, 0]]

output_wo_mask_fill = model(input_ids=input_ids, attention_mask=attention_mask, position_ids=position_ids)

input_ids = torch.masked_fill(input_ids, ~attention_mask, 0)  # [[0, 0, 0, 1, 2, 3, 4, 0]]
output_w_mask_fill = model(input_ids=input_ids, attention_mask=attention_mask, position_ids=position_ids)

print(torch.mean(torch.abs(output_wo_mask_fill.logits - output_w_mask_fill.logits), dim=-1)) # [[0.8371, 0.8371, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 3.6457]]
```

This operation is not absolutely necessary, since invalid logits are then masked:

https://github.com/huggingface/trl/blob/a67f2143c38d6520be8735463ce715ad5c281db8/trl/trainer/rloo_trainer.py#L413-L415

**Comment by Chios-C** [comment]
Thanks for your great response.
